### ISSUE TITLE
DWRF array reader E2E test failure

### ISSUE THREAD

**Issue by yingsu00** [facebookincubator/velox#3590]
### Bug description

E2EFilterTest.listAndMap fails with the following error on main branch:

/Users/yingsu/repo/velox5/velox/cmake-build-debug/velox/dwio/dwrf/test/velox_dwrf_e2e_filter_test --gtest_filter=E2EFilterTest.listAndMap:E2EFilterTest/*.listAndMap:*/E2EFilterTest.listAndMap/*:*/E2EFilterTest/*.listAndMap --gtest_color=no
Testing started at 12:04 AM ...
E20221222 00:05:29.654161 22831952 Exceptions.h:68] Line: /Users/yingsu/repo/velox5/velox/velox/dwio/common/SelectiveColumnReader.cpp:377, Function:addSkippedParentNulls, Expression: parentNullsRecordedTo_ == from (21230 vs. 22517), Source: RUNTIME, ErrorCode: INVALID_STATE
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (21230 vs. 22517)
Retriable: False
Expression: parentNullsRecordedTo_ == from
Function: addSkippedParentNulls
File: /Users/yingsu/repo/velox5/velox/velox/dwio/common/SelectiveColumnReader.cpp
Line: 377

### System information

N/A

**Comment by yingsu00** [comment]
cc @oerling @Yuhta

**Comment by Yuhta** [comment]
It's passing in latest version on CircleCI: https://app.circleci.com/pipelines/github/facebookincubator/velox/18100/workflows/99375245-af73-42dc-a96f-654a5ba979db/jobs/114366/tests
Also passing on devserver, I don't see it flaky on TestX.
Maybe check if there is some change on your side that is not pushed to master yet.

**Comment by yingsu00** [comment]
@Yuhta I triple checked my environment and made sure my repo was up to date and there were no local changes both yesterday and today. I also cloned a totally new repo and rebuilt from scratch. The error remains.

**Comment by Yuhta** [comment]
Are you on ARM or any special hardware?

**Comment by yingsu00** [comment]
No. Im with Intel cpu. I cloned the new repo, ran setup script, and build from scratch. Then I run the tests. All other tests passed except listAndMap in E2EFilterTest

**Comment by Yuhta** [comment]
Can you dump the data in `sinkPtr_` to a file and attach it here?

**Comment by yingsu00** [comment]
@Yuhta Jimmy please see the attached DWRF file
[listAndMap_NoRGSkip.dwrf 2.zip](https://github.com/facebookincubator/velox/files/10372064/listAndMap_NoRGSkip.dwrf.2.zip)

**Comment by Yuhta** [comment]
@yingsu00 I tried reading the file and run test on it and still got no error.  When does the error happen, is it in `readWithoutFilter` or `readWithFilter`?  If the latter, what is the filter?

FYI the code I use:
```c++
TEST_F(E2EFilterTest, listAndMapCrash) {
  auto file = std::make_shared<LocalReadFile>(
      getExampleFilePath("listAndMap_NoRGSkip.dwrf"));
  ReaderOptions readerOptions;
  auto reader = DwrfReader::create(
      std::make_unique<BufferedInput>(file, *pool_), readerOptions);
  RowReaderOptions rowReaderOptions;
  auto rowReader = reader->createRowReader(rowReaderOptions);
  std::vector<RowVectorPtr> batches;
  for (;;) {
    VectorPtr vector;
    if (rowReader->next(1024, vector) == 0) {
      break;
    }
    batches.push_back(std::static_pointer_cast<RowVector>(vector));
  }
  constexpr const char* kColumns =
      "long_val:bigint,"
      "long_val_2:bigint,"
      "int_val:int,"
      "array_val:array<struct<array_member: array<int>>>,"
      "map_val:map<bigint,struct<nested_map: map<int, int>>>";
  rowType_ = test::DataSetBuilder::makeRowType(kColumns, true);
  filterGenerator_ = std::make_unique<FilterGenerator>(rowType_, 1);
  DataBuffer<char> buf(*pool_, file->size());
  file->pread(0, file->size(), buf.data());
  MemorySink sink(*pool_, file->size());
  sink.write(std::move(buf));
  sinkPtr_ = &sink;
  testNoRowGroupSkip(batches, {"long_val", "long_val_2", "int_val"}, 10);
}
```
